source/CIS/cis-5.rst

```diff
@@ -0,0 +1,599 @@
+            | (1: Byte) (address: ContractAddress)
+
+
+.. _CIS-5-EntrypointName:
+
+``EntrypointName``
+^^^^^^^^^^^^^^^^^^
+
+A name for a smart contract function entrypoint.
+
+It is serialized as: First 2 bytes encode the length (``n``) of the entrypoint name in little-endian, followed by this many bytes for the entrypoint name (``entrypoint``)::
+
+  EntrypointName ::= (n: Byte²) (entrypoint: Byteⁿ)
+
+.. _CIS-5-Timestamp:
+
+``Timestamp``
+^^^^^^^^^^^^^
+
+A timestamp given in milliseconds since Unix epoch.
+It consists of an unsigned 64-bit integer.
+
+It is serialized as 8 bytes in little-endian::
+
+  Timestamp ::= (milliseconds: Byte⁸)
+
+.. _CIS-5-Nonce:
+
+``Nonce``
+^^^^^^^^^
+
+An unsigned 64-bit integer number that increases sequentially to protect against replay attacks.
+
+It is serialized as 8 bytes in little-endian::
+
+  Nonce ::= (nonce: Byte⁸)
+
+.. _CIS-5-CCDAmount:
+
+``CCDAmount``
+^^^^^^^^^^^^^
+
+An unsigned 64-bit integer number.
+
+It is serialized as 8 bytes in little-endian::
+
+  CCDAmount ::= (ccdAmount: Byte⁸)
+
+.. _CIS-5-PublicKeyEd25519:
+
+``PublicKeyEd25519``
+^^^^^^^^^^^^^^^^^^^^
+
+An Ed25519 public key is represented as a 32-byte array.
+
+It is serialized as 32 bytes::
+
+  PublicKeyEd25519 ::= (key: Byte³²)
+
+.. _CIS-5-SignatureEd25519:
+
+``SignatureEd25519``
+^^^^^^^^^^^^^^^^^^^^
+
+Signature for an Ed25519 message.
+
+It is serialized as 64 bytes::
+
+  SignatureEd25519 ::= (signature: Byte⁶⁴)
+
+.. _CIS-5-SigningData:
+
+``SigningData``
+^^^^^^^^^^^^^^^
+
+Signing data contains metadata for the signature that is used to check whether the signed message is designated for the correct contract and entrypoint, and that it is not expired.
+
+It is serialized as :ref:`CIS-5-ContractAddress` (``contract_address``), :ref:`CIS-5-EntrypointName` (``entrypoint``), :ref:`CIS-5-Nonce` (``nonce``), :ref:`CIS-5-Timestamp` (``timestamp``), :ref:`CIS-5-CCDAmount`/:ref:`CIS-5-TokenAmount` (``serviceFee``), and :ref:`CIS-5-Address` (``serviceFeeRecipient``)::
+
+  SigningData ::= (contract_address: ContractAddress) (entrypoint: EntrypointName) (nonce: Nonce) (timestamp: Timestamp) (serviceFee: CCDAmount/tokenAmount) (serviceFeeRecipient: Address)
+
+For each of the signature checking endpoints the signing data is as follows::
+
+  WithdrawNativeCurrencySigningData ::= (to: Receiver) (ccdAmount: CCDAmount) (data: AdditionalData) (signingData: SigningData)
+
+  WithdrawCis2TokensSigningData ::= (to: Receiver) (tokenAmount: tokenAmount) (tokenId: tokenID) (cis2TokenContractAddress: ContractAddress) (data: AdditionalData) (signingData: SigningData)
+
+  InternalNativeCurrencyTransferSigningData ::= (from: PublicKeyEd25519) (to: PublicKeyEd25519) (ccdAmount: CCDAmount) (signingData: SigningData)
+
+  InternalCis2TokensTransferSigningData ::= (from: PublicKeyEd25519) (to: PublicKeyEd25519) (tokenAmount: tokenAmount) (tokenID: TokenID) (cis2TokenContractAddress: ContractAddress) (signingData: SigningData)
```

Comment:
**Question:** I can't seem to find where these definitions are used, are they needed?

Reply:
If we add support for CIS-5 into directly into the wallets then we need to defined in the standard the exact serialization of the message to be signed.

In the way CIS-2 transfer is implemented in the Concordium wallet directly, the `transfer/withdraw/deposit` from a CIS-5 smartContractWallet will be directly integrated into the Concordium wallet. At least that is my understanding of support CIS-5 as first class citizens in our wallets.

Reply:
If we add support for CIS-5 directly into the wallets then we need to defined in the standard the exact serialization of the message to be signed.

In the way CIS-2 transfer is implemented in the Concordium wallet directly, the `transfer/withdraw/deposit` from a CIS-5 smartContractWallet will be directly integrated into the Concordium wallet. At least that is my understanding of supporting the CIS-5 contracts as first class citizens in our wallets.

Reply:
If we add support for CIS-5 directly into the wallets then we need to defined in the standard the exact serialization of the message to be signed.

In the way CIS-2 transfer is implemented in the Concordium wallet directly, the `transfer/withdraw/deposit` from a CIS-5 smartContractWallet will be directly integrated into the Concordium wallet. At least that is my understanding of supporting the CIS-5 contracts as first class citizens in our wallets.